File: librats/nodejs/examples/basic_client.js
const { RatsClient, ConnectionStrategy } = require('../lib/index');

/**
 * Basic client example demonstrating connection and message exchange
 */
class BasicClientExample {
  constructor(port = 8080) {
    this.client = new RatsClient(port);
    this.setupCallbacks();
  }

  setupCallbacks() {
    // Set up event handlers
    this.client.onConnection((peerId) => {
      console.log(`✅ Peer connected: ${peerId}`);
      
      // Send a welcome message to the new peer
      this.client.sendString(peerId, "Hello! Welcome to the network.");
    });

    this.client.onString((peerId, message) => {
      console.log(`📝 String message from ${peerId}: ${message}`);
      
      // Echo the message back
      this.client.sendString(peerId, `Echo: ${message}`);
    });

    this.client.onBinary((peerId, data) => {
      console.log(`📦 Binary message from ${peerId}, size: ${data.length} bytes`);
      console.log(`📦 Data: ${data.toString('hex')}`);
    });

    this.client.onJson((peerId, jsonStr) => {
      try {
        const data = JSON.parse(jsonStr);
        console.log(`🔧 JSON message from ${peerId}:`, data);
      } catch (e) {
        console.log(`❌ Invalid JSON from ${peerId}: ${jsonStr}`);
      }
    });

    this.client.onDisconnect((peerId) => {
      console.log(`❌ Peer disconnected: ${peerId}`);
    });
  }

  async start() {
    console.log('🚀 Starting RatsClient...');
    
    if (!this.client.start()) {
      throw new Error('Failed to start client');
    }

    console.log(`✅ Client started successfully`);
    console.log(`📋 Our peer ID: ${this.client.getOurPeerId()}`);
    
    // Enable encryption for secure communication
    this.client.setEncryptionEnabled(true);
    const encKey = this.client.generateEncryptionKey();
    console.log(`🔐 Generated encryption key: ${encKey}`);
  }

  connectToPeer(host, port, strategy = ConnectionStrategy.AUTO_ADAPTIVE) {
    console.log(`🔗 Connecting to ${host}:${port} using strategy ${strategy}`);
    
    if (this.client.connectWithStrategy(host, port, strategy)) {
      console.log(`✅ Connection initiated successfully`);
    } else {
      console.log(`❌ Failed to initiate connection`);
    }
  }

  sendTestMessages() {
    const peerIds = this.client.getPeerIds();
    
    if (peerIds.length === 0) {
      console.log('No peers connected to send messages to');
      return;
    }

    console.log(`📤 Sending test messages to ${peerIds.length} peer(s)`);

    // Send different types of messages
    peerIds.forEach(peerId => {
      // String message
      this.client.sendString(peerId, `Hello from ${this.client.getOurPeerId()}!`);
      
      // Binary message
      const binaryData = Buffer.from('Hello World!', 'utf8');
      this.client.sendBinary(peerId, binaryData);
      
      // JSON message
      const jsonData = {
        type: 'greeting',
        from: this.client.getOurPeerId(),
        timestamp: Date.now(),
        message: 'Hello from Node.js!'
      };
      this.client.sendJson(peerId, JSON.stringify(jsonData));
    });
  }

  broadcastTestMessages() {
    console.log('📡 Broadcasting test messages...');
    
    // Broadcast string
    const stringCount = this.client.broadcastString('Broadcast message from Node.js!');
    console.log(`📤 String broadcast sent to ${stringCount} peers`);
    
    // Broadcast binary
    const binaryData = Buffer.from('Binary broadcast data', 'utf8');
    const binaryCount = this.client.broadcastBinary(binaryData);
    console.log(`📤 Binary broadcast sent to ${binaryCount} peers`);
    
    // Broadcast JSON
    const jsonData = {
      type: 'broadcast',
      from: this.client.getOurPeerId(),
      timestamp: Date.now(),
      message: 'This is a JSON broadcast'
    };
    const jsonCount = this.client.broadcastJson(JSON.stringify(jsonData));
    console.log(`📤 JSON broadcast sent to ${jsonCount} peers`);
  }

  printStatus() {
    console.log('\n📊 Client Status:');
    console.log(`   Peer Count: ${this.client.getPeerCount()}`);
    console.log(`   Our Peer ID: ${this.client.getOurPeerId()}`);
    console.log(`   Max Peers: ${this.client.getMaxPeers()}`);
    console.log(`   Peer Limit Reached: ${this.client.isPeerLimitReached()}`);
    console.log(`   Encryption Enabled: ${this.client.isEncryptionEnabled()}`);
    
    const stats = this.client.getConnectionStatistics();
    if (stats) {
      console.log(`   Connection Statistics: ${stats}`);
    }
    
    const peerIds = this.client.getPeerIds();
    if (peerIds.length > 0) {
      console.log(`   Connected Peers: ${peerIds.join(', ')}`);
    }
    console.log('');
  }

  stop() {
    console.log('🛑 Stopping client...');
    this.client.stop();
  }
}

// Example usage
async function main() {
  const args = process.argv.slice(2);
  const port = args[0] ? parseInt(args[0]) : 8080;
  
  const client = new BasicClientExample(port);
  
  try {
    await client.start();
    
    // Print status every 10 seconds
    const statusInterval = setInterval(() => {
      client.printStatus();
    }, 10000);
    
    // If host and port are provided, connect to a peer
    if (args.length >= 2) {
      const host = args[1];
      const peerPort = parseInt(args[2]) || 8081;
      
      setTimeout(() => {
        client.connectToPeer(host, peerPort);
      }, 1000);
    }
    
    // Send test messages every 15 seconds if we have peers
    const messageInterval = setInterval(() => {
      if (client.client.getPeerCount() > 0) {
        client.sendTestMessages();
      }
    }, 15000);
    
    // Broadcast test messages every 30 seconds if we have peers
    const broadcastInterval = setInterval(() => {
      if (client.client.getPeerCount() > 0) {
        client.broadcastTestMessages();
      }
    }, 30000);
    
    // Handle graceful shutdown
    process.on('SIGINT', () => {
      console.log('\n🛑 Received SIGINT, shutting down gracefully...');
      clearInterval(statusInterval);
      clearInterval(messageInterval);
      clearInterval(broadcastInterval);
      client.stop();
      process.exit(0);
    });
    
    console.log('✅ Client is running. Press Ctrl+C to stop.');
    console.log('💡 Usage: node basic_client.js [listen_port] [connect_host] [connect_port]');
    
  } catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = BasicClientExample;

File: librats/python/librats_py/examples/basic_client.py
#!/usr/bin/env python3
"""
Basic librats client example.

This example demonstrates basic peer-to-peer communication using librats.
"""

import time
import sys
import threading
from librats_py import RatsClient, RatsError


def on_peer_connected(peer_id: str):
    """Called when a new peer connects."""
    print(f"✓ Peer connected: {peer_id}")


def on_peer_disconnected(peer_id: str):
    """Called when a peer disconnects."""
    print(f"✗ Peer disconnected: {peer_id}")


def on_string_message(peer_id: str, message: str):
    """Called when a string message is received."""
    print(f"📨 Message from {peer_id}: {message}")


def on_json_message(peer_id: str, data: dict):
    """Called when a JSON message is received."""
    print(f"📨 JSON from {peer_id}: {data}")


def main():
    if len(sys.argv) != 2:
        print("Usage: python basic_client.py <listen_port>")
        print("Example: python basic_client.py 8080")
        sys.exit(1)
    
    try:
        listen_port = int(sys.argv[1])
    except ValueError:
        print("Error: Port must be a number")
        sys.exit(1)
    
    # Create and configure the client
    try:
        with RatsClient(listen_port) as client:
            print(f"🚀 Starting librats client on port {listen_port}")
            
            # Set up callbacks
            client.set_connection_callback(on_peer_connected)
            client.set_disconnect_callback(on_peer_disconnected)
            client.set_string_callback(on_string_message)
            client.set_json_callback(on_json_message)
            
            # Enable logging
            RatsClient.set_logging_enabled(True)
            
            # Start the client
            client.start()
            print(f"✓ Client started successfully")
            print(f"🆔 Our peer ID: {client.get_our_peer_id()}")
            
            # Interactive command loop
            print("\nCommands:")
            print("  connect <host> <port>  - Connect to a peer")
            print("  send <peer_id> <msg>   - Send a message to a peer")
            print("  broadcast <message>    - Broadcast message to all peers")
            print("  peers                  - Show connected peers")
            print("  stats                  - Show connection statistics")
            print("  quit                   - Exit the program")
            print()
            
            while True:
                try:
                    command = input("librats> ").strip().split()
                    if not command:
                        continue
                    
                    cmd = command[0].lower()
                    
                    if cmd == "quit" or cmd == "exit":
                        break
                    
                    elif cmd == "connect":
                        if len(command) != 3:
                            print("Usage: connect <host> <port>")
                            continue
                        
                        try:
                            host = command[1]
                            port = int(command[2])
                            client.connect(host, port)
                            print(f"🔗 Connecting to {host}:{port}...")
                        except ValueError:
                            print("Error: Port must be a number")
                        except RatsError as e:
                            print(f"Connection failed: {e}")
                    
                    elif cmd == "send":
                        if len(command) < 3:
                            print("Usage: send <peer_id> <message>")
                            continue
                        
                        peer_id = command[1]
                        message = " ".join(command[2:])
                        
                        try:
                            client.send_string(peer_id, message)
                            print(f"📤 Sent message to {peer_id}")
                        except RatsError as e:
                            print(f"Send failed: {e}")
                    
                    elif cmd == "broadcast":
                        if len(command) < 2:
                            print("Usage: broadcast <message>")
                            continue
                        
                        message = " ".join(command[1:])
                        count = client.broadcast_string(message)
                        print(f"📡 Broadcasted message to {count} peers")
                    
                    elif cmd == "peers":
                        count = client.get_peer_count()
                        print(f"👥 Connected peers: {count}")
                        
                        # Show connection statistics
                        stats = client.get_connection_statistics()
                        if stats and 'peers' in stats:
                            for peer_info in stats['peers']:
                                print(f"  - {peer_info}")
                    
                    elif cmd == "stats":
                        stats = client.get_connection_statistics()
                        print("📊 Connection Statistics:")
                        for key, value in stats.items():
                            print(f"  {key}: {value}")
                    
                    else:
                        print(f"Unknown command: {cmd}")
                
                except KeyboardInterrupt:
                    break
                except EOFError:
                    break
                except Exception as e:
                    print(f"Error: {e}")
            
            print("\n👋 Shutting down...")
    
    except RatsError as e:
        print(f"❌ Error: {e}")
        sys.exit(1)
    except KeyboardInterrupt:
        print("\n👋 Interrupted by user")
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
